File: golang/src/_53_maximum_subarray.go
/*
 * @lc app=leetcode id=53 lang=golang
 *
 * [53] Maximum Subarray
 */

// @lc code=start
func maxSubArray(nums []int) int {
	maxSum := nums[0]
	curSum := nums[0]
	for _, n := range nums[1:] {
		if curSum < 0 {
			curSum = n
		} else {
			curSum += n
		}
		if curSum > maxSum {
			maxSum = curSum
		}
	}
	return maxSum
}

// @lc code=end

File: kotlin/src/_53_maximum_subarray.kt
/*
 * @lc app=leetcode id=53 lang=kotlin
 *
 * [53] Maximum Subarray
 */

// @lc code=start
class Solution {
    fun maxSubArray(nums: IntArray): Int {
        var maxSum = nums[0]
        var curSum = nums[0]
        for (i in 1 until nums.size) {
            curSum = maxOf(nums[i], curSum + nums[i])
            maxSum = maxOf(maxSum, curSum)
        }
        return maxSum
    }
}
// @lc code=end
